### PR TITLE
Add `option_layer`

### PR DESCRIPTION
**Pull request by davidpdrsn**
Tower already has `option_layer` for conditionally applying a `Layer` but since it uses `tower::util::Either` it always changes the error type to `BoxError`. That requires using `HandleErrorLayer` which is inconvenient for axum users.

That has been changed in tower
(https://github.com/tower-rs/tower/pull/637) but still has some issues (https://github.com/tower-rs/tower/issues/665). Its also a breaking change so hasn't been released yet.

In the meantime I figure we can provide our own thing in axum-extra, since we already have an `Either` type there.

### CODE REVIEW

File: axum-extra/src/middleware.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+/// use axum::{Router, routing::get};
+/// use std::time::Duration;
+/// use tower_http::timeout::TimeoutLayer;
+///
+/// # let option_timeout = Some(Duration::new(10, 0));
+/// let timeout_layer = option_timeout.map(TimeoutLayer::new);
+///
+/// let app = Router::new()
+///     .route("/", get(|| async {}))
+///     .layer(option_layer(timeout_layer));
```
Comment (Oliboy50) [inline axum-extra/src/middleware.rs:23]: Just to be sure, if we pass `.layer(Identity::new())`, it means zero additional cost for each incoming request?
Just like we wouldn't write `.layer(...)` at all?

**Review by Oliboy50 — APPROVED** [review]
🙇

File: axum-extra/src/middleware.rs (hunk 1 above)
Comment (davidpdrsn) [reply axum-extra/src/middleware.rs:23]: It does have additional cost. It'll create a new [`Route`](https://github.com/tokio-rs/axum/blob/main/axum/src/routing/route.rs#L30) which requires allocation and a dynamic dispatch. I doubt that impact will be measurable in a real world app however but if you really wanna avoid it you have to use an `if`.

Comment (davidpdrsn) [reply axum-extra/src/middleware.rs:23]: And by the way this is why we recommend adding multiple middleware with [`ServiceBuilder`](https://docs.rs/axum/latest/axum/middleware/index.html#applying-multiple-middleware).

**Review by jplatte — APPROVED** [review]

**Pull request merged**